### PR TITLE
feat: load source data to redshift tables

### PR DESCRIPTION
**Pull request by amliuyong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary
- load transformed event files from S3 to redshift event table
- load transformed event_parameter files from S3 to redshift event_parameter table
- load transformed user files  from S3 to redshift user table
- load transformed item files from S3 to redshift item table
- refactor sp_clear_*.sql
- add new MATERIALIZED views: user_m_view and item_m_view

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [x] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [x] deploy data processing
  - [ ] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/analytics/private/metrics-common-workflow.ts [hunk 1]
```diff
@@ -105,96 +120,60 @@ export function buildMetricsWidgetForWorkflows(scope: Construct, id: string, pro
...
-        title: 'Clear expired events execution time',
-        view: 'timeSeries',
-        metrics: [
-          [statesNamespace, 'ExecutionTime', ...clearExpiredEventsWorkflowDimension],
-        ],
+      {
+        type: 'metric',
+        properties: {
+          stat: 'Average',
+          title: `Load ${dimName[1]}  workflow execution time`,
```
Comment (zxkane) [inline src/analytics/private/metrics-common-workflow.ts:166]: Two spaces before `workflow`

Comment (zxkane) [reply src/analytics/private/metrics-common-workflow.ts:166]: Add `'` quotes for dimension name?

File: src/analytics/private/sqls/redshift/event-parameter.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS {{schema}}.{{table_event_parameter}}(
+    event_timestamp BIGINT,
+    event_id VARCHAR(255) DEFAULT RANDOM(),
+    event_param_key VARCHAR(255),
+    event_param_double_value DOUBLE PRECISION,
+    event_param_float_value DOUBLE PRECISION,
+    event_param_int_value BIGINT,
+    event_param_string_value VARCHAR(255)    
+) DISTSTYLE EVEN 
+SORTKEY(event_id)
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/event-parameter.sql:10]: id is a random string, how to use it for optimizing the query?

@qiaow02 @luorobin-a2z any suggestion?

File: src/analytics/private/sqls/redshift/sp-clear-item-and-user.sql [hunk 3]
```diff
@@ -0,0 +1,37 @@
+CREATE OR REPLACE PROCEDURE {{schema}}.sp_clear_item_and_user()
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-clear-item-and-user.sql:1]: What's the logic to clear the legacy items and users?

File: src/analytics/private/sqls/redshift/event-parameter.sql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS {{schema}}.{{table_event_parameter}}(
+    event_timestamp BIGINT,
+    event_id VARCHAR(255) DEFAULT RANDOM(),
+    event_param_key VARCHAR(255),
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/event-parameter.sql:4]: should we add a field for the type of event value?

Comment (zxkane) [reply src/analytics/private/sqls/redshift/event-parameter.sql:4]: And how do we handle with the same event parameter name in different events?

File: test/utils.ts [hunk 5]
```diff
@@ -25,6 +25,10 @@ export const RefAnyValue = {
   Ref: Match.anyValue(),
 };
 
+export const JonAnyValue = {
```
Comment (zxkane) [inline test/utils.ts:28]: to `JoinAnyValue`

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/parameter.ts [hunk 6]
```diff
@@ -184,13 +184,13 @@ export function createStackParameters(scope: Construct): {
 
   const { projectIdParam, appIdsParam } = Parameters.createProjectAndAppsParameters(scope, 'ProjectId', 'AppIds');
 
-  const odsEventBucketParam = Parameters.createS3BucketParameter(scope, 'ODSEventBucket', {
-    description: 'The S3 bucket name for ODS events.',
+  const sourceBucketParam = Parameters.createS3BucketParameter(scope, 'SourceDataBucket', {
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/analytics/parameter.ts:187]: The updated logic needs to be modified. 
- The original logic was to **read** the workflow JSON from DDB and modify the corresponding **parameters**. 
- It needs to be modified to: **regenerate** the workflow JSON based on the new parameters

Refer to the generation logic when **creating** a new pipeline.

[the remark this answers is not in the document]
Comment (zxkane) [reply src/analytics/parameter.ts:187]: Re-using existing parameters as base parameters make senses for updating. For upgrading, we need build the complete parameters based on latest code base.

File: src/analytics/private/app-schema.ts [hunk 7]
```diff
@@ -128,7 +129,7 @@ export class ApplicationSchemas extends Construct {
       handler: 'handler',
       memorySize: 128,
       reservedConcurrentExecutions: 1,
-      timeout: Duration.minutes(5),
+      timeout: Duration.minutes(15),
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/private/app-schema.ts:132]: increase it for test, changed back.

File: src/analytics/private/load-ods-data-workflow.ts [hunk 8]
```diff
@@ -50,25 +48,24 @@ export interface LoadODSEventToRedshiftWorkflowProps {
...
+  public readonly loadDataWorkflow: IStateMachine;
 
-  public readonly loadEventWorkflow: IStateMachine;
-
-  constructor(scope: Construct, id: string, props: LoadODSEventToRedshiftWorkflowProps) {
+  constructor(scope: Construct, id: string, props: LoadOdsDataToRedshiftWorkflowProps) {
     super(scope, id);
 
-    const odsEventTable = this.odsEventTable();
+    const ddbStatusTable = this.createDDBStatusTable(props.odsTableName);
```
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/private/load-ods-data-workflow.ts:62]: changed to use only one ddb table. set jobStatus to pattern"<ods_table>#status" to determine the specific destination table.  e.g. "event_parameter#JOB_NEW"

File: test/utils.ts (hunk 5 above)
Comment (amliuyong) [reply test/utils.ts:28]: changed.

File: src/analytics/private/sqls/redshift/event-parameter.sql (hunk 4 above)
Comment (amliuyong) [reply src/analytics/private/sqls/redshift/event-parameter.sql:4]: added `event_name` to table.  the `event_id` which is a foreign key refer to event table. the `DEFAULT RANDOM()` here made the confusion, removed it.

File: src/analytics/private/sqls/redshift/sp-clear-item-and-user.sql (hunk 3 above)
Comment (amliuyong) [reply src/analytics/private/sqls/redshift/sp-clear-item-and-user.sql:1]: remove older user/item for the same id, keep the latest one based on `event_timestamp`

File: src/analytics/private/sqls/redshift/event-parameter.sql (hunk 2 above)
Comment (amliuyong) [reply src/analytics/private/sqls/redshift/event-parameter.sql:10]: add event_name to this table, add changed the `SORTKEY(event_timestamp, event_name)`, make the SORTKEY same as `event` table, we can get better performance when do join with event.

File: src/analytics/private/metrics-common-workflow.ts (hunk 1 above)
Comment (amliuyong) [reply src/analytics/private/metrics-common-workflow.ts:166]: changed

File: src/analytics/parameter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (amliuyong) [reply src/analytics/parameter.ts]: changed to re-using existing parameters.  now no breaking changes are introduced. 
also add a customer resource `GetSourcePrefixCustomerResource` to handle the prefix.

File: src/analytics/lambdas/custom-resource/get-source-prefix.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+async function _handler(event: CdkCustomResourceEvent) {
+  const requestType = event.RequestType;
+  logger.info('RequestType: ' + requestType);
+
+  const odsEventPrefix = (event.ResourceProperties as PropertiesType).odsEventPrefix;
+
+  let prefix = odsEventPrefix;
+
+  if (odsEventPrefix.endsWith('/ods_events/') || odsEventPrefix.endsWith('/ods_events')) {
+    prefix = odsEventPrefix.replace(new RegExp('/ods_events[/]?'), '');
```
Comment (zxkane) [inline src/analytics/lambdas/custom-resource/get-source-prefix.ts:52]: Now you removed `ods_events` in prefix, what's the final prefix looking like?

File: src/analytics/lambdas/load-data-workflow/put-ods-source-to-store.ts [hunk 10]
```diff
@@ -93,6 +94,8 @@ export const handler = async (event: EventBridgeEvent<'Object Created', S3Object
  * @param timestamp The S3 object create event notification timestamp.
  */
 export const putItem = async (tableName: string, s3Bucket: string, s3Object: string, s3ObjSize: string, jobStatus: string, timestamp: number) => {
+  const qJobStatus = composeJobStatus(jobStatus, REDSHIFT_ODS_TABLE_NAME);
```
Comment (zxkane) [inline src/analytics/lambdas/load-data-workflow/put-ods-source-to-store.ts:97]: Is this change covered by test case?

File: src/analytics/private/metrics-common-workflow.ts [hunk 11]
```diff
@@ -58,19 +74,18 @@ export function buildMetricsWidgetForWorkflows(scope: Construct, id: string, pro
...
-  const loadEventsWorkflowAlarm = new Alarm(scope, id + 'loadEventsWorkflowAlarm', {
+  const loadEventsWorkflowAlarm = new Alarm(scope, id + 'loadEventWorkflowAlarm', {
     comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
     threshold: 1,
     evaluationPeriods: 1,
     treatMissingData: TreatMissingData.NOT_BREACHING,
-    metric: props.loadEventsWorkflow.metricFailed({ period: Duration.hours(1) }),
-    alarmDescription: `Load events workflow failed, projectId: ${props.projectId}`,
-    alarmName: getAlarmName(scope, props.projectId, 'Load events workflow'),
+    metric: props.loadDataWorkflows.event.metricFailed({ period: Duration.hours(1) }),
```
Comment (zxkane) [inline src/analytics/private/metrics-common-workflow.ts:82]: Should the duration aligh with the interval of running data processing?

File: src/analytics/private/metrics-common-workflow.ts [hunk 12]
```diff
@@ -58,19 +74,18 @@ export function buildMetricsWidgetForWorkflows(scope: Construct, id: string, pro
     'service', MetricsService.WORKFLOW,
   ];
 
-  const loadEventsWorkflowAlarm = new Alarm(scope, id + 'loadEventsWorkflowAlarm', {
+  const loadEventsWorkflowAlarm = new Alarm(scope, id + 'loadEventWorkflowAlarm', {
```
Comment (zxkane) [inline src/analytics/private/metrics-common-workflow.ts:77]: Should we rename it to `LoadDataWorkflowAlarm`?

Just see a single alarm here, do we share the workflow to load data to tables?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/analytics/lambdas/load-data-workflow/put-ods-source-to-store.ts (hunk 10 above)
Comment (amliuyong) [reply src/analytics/lambdas/load-data-workflow/put-ods-source-to-store.ts:97]: yes.

File: src/analytics/lambdas/custom-resource/get-source-prefix.ts (hunk 9 above)
Comment (amliuyong) [reply src/analytics/lambdas/custom-resource/get-source-prefix.ts:52]: below are the prefixes for each table:

 - xxxx/project_id/ods_events/
 - xxxx/project_id/event/
 - xxxx/project_id/event_parameter/
 - xxxx/project_id/user/
 - xxxx/project_id/item/

File: src/analytics/private/metrics-common-workflow.ts (hunk 12 above)
Comment (amliuyong) [reply src/analytics/private/metrics-common-workflow.ts:77]: no. each destination table has a workflow, there are 4 workflows.

File: src/analytics/private/metrics-common-workflow.ts (hunk 11 above)
Comment (amliuyong) [reply src/analytics/private/metrics-common-workflow.ts:82]: the line-87, as below, overrides the hard-code value: 
```
  (loadEventsWorkflowAlarm.node.defaultChild as CfnResource).addPropertyOverride('Period', processingJobInterval.getIntervalSeconds());

```

cannot use the customer resource value directly in line-82, the hard-code value just a place-holder.

add comments to the hard-code lines.

File: src/analytics/private/metrics-common-workflow.ts (hunk 12 above)
Comment (zxkane) [reply src/analytics/private/metrics-common-workflow.ts]: Will we have four alarms for four loading workflows?

Comment (amliuyong) [reply src/analytics/private/metrics-common-workflow.ts]: No. only 2 alarams, one for failed workflow, one for max file age.

Comment (zxkane) [reply src/analytics/private/metrics-common-workflow.ts]: Will the failure of any table loading trigger the alarm?

**Review by zxkane — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**